source/extensions/common/tap/tap_matcher.cc

```diff
@@ -0,0 +1,133 @@
+#include "extensions/common/tap/tap_matcher.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Tap {
+
+void buildMatcher(const envoy::service::tap::v2alpha::MatchPredicate& match_config,
+                  std::vector<MatcherPtr>& matchers) {
+  // In order to store indexes and build our matcher tree inline, we must reserve a slot where
+  // the matcher we are about to create will go. This allows us to know its future index and still
+  // construct more of the tree in each called constructor (e.g., multiple OR/AND conditions).
+  // Once fully constructed, we move the matcher into its position below. See the tap matcher
+  // overview in tap.h for more information.
+  matchers.emplace_back(nullptr);
+
+  MatcherPtr new_matcher;
+  switch (match_config.rule_case()) {
+  case envoy::service::tap::v2alpha::MatchPredicate::kOrMatch:
+    new_matcher = std::make_unique<SetLogicMatcher>(match_config.or_match(), matchers,
+                                                    SetLogicMatcher::Type::Or);
+    break;
+  case envoy::service::tap::v2alpha::MatchPredicate::kAndMatch:
+    new_matcher = std::make_unique<SetLogicMatcher>(match_config.and_match(), matchers,
+                                                    SetLogicMatcher::Type::And);
+    break;
+  case envoy::service::tap::v2alpha::MatchPredicate::kNotMatch:
+    new_matcher = std::make_unique<NotMatcher>(match_config.not_match(), matchers);
+    break;
+  case envoy::service::tap::v2alpha::MatchPredicate::kAnyMatch:
+    new_matcher = std::make_unique<AnyMatcher>(matchers);
+    break;
+  case envoy::service::tap::v2alpha::MatchPredicate::kHttpRequestMatch:
+    new_matcher = std::make_unique<HttpRequestMatcher>(match_config.http_request_match(), matchers);
+    break;
+  case envoy::service::tap::v2alpha::MatchPredicate::kHttpResponseMatch:
+    new_matcher =
+        std::make_unique<HttpResponseMatcher>(match_config.http_response_match(), matchers);
+    break;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+
+  // Per above, move the matcher into its position.
+  matchers[new_matcher->index()] = std::move(new_matcher);
```

Comment:
How come we need the vector inlining? Can't we just build a regular tree and maintain a pointer to the current node of interest, tracking the leaf-parent path booleans as a vector if necessary?

Reply:
Sorry I don't follow what you are suggesting. This implementation allows a per-request status vector to be created and then trivially updated without worrying about pointers. IMO it's both the simplest and most efficient implementation for what we need to do.

Reply:
Fair enough, I don't disagree that it's an efficient implementation. I'm visualizing this as at any point in time, for a given request, you have the match tree (immutable and shared) and your current status, which is those parts that have matched. I can see the attractiveness of the flattening due to the fact that you can easily map either way between match predicate and status. I